File: src/common/Position3D.cpp
/*
**********************************************************************************

 © 2019 Arizona Board of Regents on behalf of the University of Arizona with rights
       granted for USDOT OSADP distribution with the Apache 2.0 open source license.

**********************************************************************************

  Position3D.cpp  
  Created by: Niraj Vasant Altekar
  University of Arizona   
  College of Engineering

  This code was developed under the supervision of Professor Larry Head
  in the Systems and Industrial Engineering Department.

  Revision History:
  1. 
*/
#include <iostream>
#include <Position3D.h>

using std::cout;
using std::endl;
using std::string;

// Constructor
Position3D::Position3D()
{

}

//Setters:
void Position3D::setLatitude_decimalDegree(double vehLatitude_DecimalDegree)
{
    if(vehLatitude_DecimalDegree >= VALID_LATITUDEMINDEG && vehLatitude_DecimalDegree <= VALID_LATITUDEMAXDEG)
        latitude_DecimalDegree = vehLatitude_DecimalDegree;
    else
        cout << "Latitude out of range!" << endl;
}
void Position3D::setLongitude_decimalDegree(double vehlongitude_DecimalDegree)
{
    if(vehlongitude_DecimalDegree >= VALID_LONGITUDEMINDEG && vehlongitude_DecimalDegree <= VALID_LONGITUDEMAXDEG)
        longitude_DecimalDegree = vehlongitude_DecimalDegree;
    else
        cout << "Longitude out of range!" << endl;
}
void Position3D::setElevation_meter(double vehElevation_Meter)
{
    if(vehElevation_Meter >= VALID_ELEVATIONMINMETER && vehElevation_Meter <= VALID_ELEVATIONMAXMETER)
        elevation_Meter = vehElevation_Meter;
    else if (vehElevation_Meter == UNKNOWN_ELEVATIONMETER)
        {
            elevation_Meter = vehElevation_Meter;
            cout << "Elevation unknown!" << endl;
        }
    else if (vehElevation_Meter < UNKNOWN_ELEVATIONMETER)
        elevation_Meter = VALID_ELEVATIONMINMETER;
    
    else if (vehElevation_Meter >= VALID_ELEVATIONMAXMETER)
        elevation_Meter = VALID_ELEVATIONMAXMETER;
}

//Getters:
double Position3D::getLatitude_DecimalDegree()
{
    return latitude_DecimalDegree;
}
double Position3D::getLongitude_DecimalDegree()
{
    return longitude_DecimalDegree;
}
double Position3D::getElevation_Meter()
{
    return elevation_Meter;
}

//Destructor:
Position3D::~Position3D()
{

}

File: src/vsp/hmi-controller/Position3D.py
"""
**********************************************************************************
 © 2019 Arizona Board of Regents on behalf of the University of Arizona with rights
       granted for USDOT OSADP distribution with the Apache 2.0 open source license.
**********************************************************************************
  cv-work-zone-controller.py
  Created by: Larry Head
  University of Arizona   
  College of Engineering

The purpose of the Position3D class is to represent a position on the earth in WGS-84 GPS coordinates 
by logitude, latitude, and elevation. 

Future enhancements incude:
    - computing the difference in position between two points 
    - checking bounds on position to ensure a proper translations
    - projection on to a local plane base on a reference GPS location


"""

import json

class Position3D:
    def __init__(self, latitude, longitude, elevation):
        self.latitude_DecimalDegree = latitude
        self.longitude_DecimalDegree = longitude
        self.elevation_Meter = elevation
        
    def Position3D2json(self):
        return json.dumps({"Position3D": { "latitude": self.latitude_DecimalDegree, 
                                "longitude": self.longitude_DecimalDegree, 
                                "elevation": self.elevation_Meter}}, sort_keys=False, indent = 4)

    def json2Position3D(self, json3Dposition):
        positionDict = json.loads(json3Dposition)
        self.latitude_DecimalDegree = positionDict['Position3D']['latitude']
        self.longitude_DecimalDegree = positionDict['Position3D']['longitude']
        self.elevation_DecimalDegree = positionDict['Position3D']['elevation']
        return self


     

if __name__ == '__main__':
    # simple self test code
    myPosition = Position3D(32.12345678, -112.12345678, 723)
    print(myPosition.latitude_DecimalDegree, myPosition.longitude_DecimalDegree, myPosition.elevation_Meter)
    
    #test Position3D to json
    myPosition_json = myPosition.Position3D2json()
    print(myPosition_json)
    
    #test jason 2 Position3D, remember to create an object first
    host = Position3D(0, 0, 0)
    host.json2Position3D(myPosition_json)
    print(host.latitude_DecimalDegree, host.longitude_DecimalDegree, host.elevation_DecimalDegree)
    
